perceval/utils/density_matrix.py

```diff
@@ -0,0 +1,363 @@
+
+            return result
+
+        return matrix_vector_multiplication
+
+    @staticmethod
+    def _bra_str(bs: BasicState):
+        return "<" + str(bs)[1:][:-1] + "|"
+
+    def to_svd(self, threshold: float = 1e-8, batch_size: int = 1):
+        """
+                gives back an SVDistribution from the density_matrix
+        """
+        if self.size < 50:  # if the matrix is small: array eigh method
+            matrix = self.mat.toarray()
+            w, v = eigh(matrix)
+            dic = {}
+            for k in range(w.shape[0]):
+                sv = array_to_statevector(v[:, k], self.reverse_index)
+                if w[k] > threshold:
+                    dic[sv] = w[k]
+        else:  # if the matrix is large: sparse eigsh method
+            val = np.array([])
+            vec = np.empty(shape=(self._size, 0), dtype=complex)
+            while (val > threshold).all():
+                deflated_operator = LinearOperator((self._size, self._size), matvec=self._deflation(self.mat, val, vec))
+                new_val, new_vec = eigsh(deflated_operator, batch_size)
+                val = np.concatenate((val, new_val))
+                vec = np.concatenate((vec, new_vec), axis=1)
+
+            dic = {}
+            for i in range(val.shape[0]):
+                if val[i] >= threshold:
+                    sv = StateVector()
+                    for j in range(len(vec)):
+                        sv += complex(vec[j][i])*self.reverse_index[j]
+                    sv.normalize()
+                    if sv.m != 0:
+                        dic[sv] = val[i]
+                else:
+                    continue
+        return SVDistribution(dic)
+
+    def __add__(self, other):
+        """
+        add two density Matrices together
+        """
+
+        if not isinstance(other, DensityMatrix):
+            raise TypeError("You can only add a Density Matrix to a Density Matrix")
+
+        if not self._m == other._m:
+            raise ValueError("You can't add Density Matrices acting on different numbers of mode")
+
+        n = max(self._n_max, other.n_max)
+
+        if n == self.n_max:
+            copy_mat = copy(other.mat)
+            copy_mat.resize(self._size, self._size)
+            new_mat = copy_mat + self.mat
+            new_index = self.index
+        else:
+            copy_mat = copy(self.mat)
+            copy_mat.resize(other._size, other._size)
+            new_mat = copy_mat + other.mat
+            new_index = other.index
+
+        return DensityMatrix(new_mat, new_index)
+
+    def __mul__(self, other):
+        """
+        Make a tensor product between a Density Matrix and a mixed state in any form
+        """
+
+        if isinstance(other, (int, float, complex)):
+            new_dm = copy(self)
+            new_dm.mat = other*new_dm.mat
+            return new_dm
+
+        return density_matrix_tensor_product(self, other)
+
+    def __rmul__(self, other):
```

Comment:
Maybe this method should called the method `__mul__` but permuting arguments to avoid code duplication (maybe you have to put more test in the __mul__ ?)
```
def __rmul__(self, other):
    return __mul__(other, self)
```

Reply:
In some cases, other will not be a density matrix, so if I call `__mul__` of an other class it will return something not implemented no ?

Reply:
In this case ok